{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68d8723f-c078-43be-b457-ef23feb8c80c",
   "metadata": {},
   "source": [
    "# The WGBS data analysis with new data\n",
    "## Overview \n",
    "In this notebook, we are going to explore how to run this module with a new dataset. These submodules provide a great framework for running a rigorous and scalable methylseq analysis, but there are some considerations that must be made in order to run this with your own data. We will walk through that process here so that hopefully, you are able to take these notebooks to your research group and use them for your own analysis. Notice that we do not give you all the answers in the code blocks, but if you get stuck, use the dropdowns for help. We will walk through it together in Tutorial 1 to run the Bismark stage of analysis. After you complete this tutorial, try to use the same logic to build out the other tutorials with this new dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49bb03da",
   "metadata": {},
   "source": [
    "## Learning Objectives\n",
    "\n",
    "* **Setting up a bioinformatics analysis environment:** Learners will practice installing necessary bioinformatics tools (FastQC, MultiQC, Bismark, Trim Galore!, Bedtools, Samtools, Metilene) using conda/mamba within a Jupyter Notebook environment.  This includes managing software versions and dependencies.\n",
    "\n",
    "* **Data Acquisition and Management:**  Learners will download a WGBS (Whole Genome Bisulfite Sequencing) dataset and reference genome from a Google Cloud Storage bucket using `gsutil` commands, organizing files into a project directory structure. This reinforces best practices for handling large datasets and reproducible research.\n",
    "\n",
    "* **WGBS Data Preprocessing:** Learners will use FastQC to assess the quality of raw sequencing data, then use Trim Galore! for adapter trimming and quality filtering of the sequencing reads. They'll learn to interpret quality control reports generated by FastQC and summarize the reports with MultiQC.\n",
    "\n",
    "\n",
    "* **Bismark Pipeline Execution:** Learners will execute the Bismark pipeline for alignment of bisulfite-converted reads to a reference genome. This includes preparing the reference genome using Bismark's genome preparation tools, aligning reads using Bismark's aligner and generating reports for analysis.\n",
    "\n",
    "\n",
    "* **Bismark Data Processing and Quality Control:** They will perform deduplication of Bismark alignments using `deduplicate_bismark`. They'll then extract methylation ratios from Bismark alignment files to generate bedGraph and coverage files using `bismark_methylation_extractor`. Learners also perform quality control checks on generated files.\n",
    "\n",
    "\n",
    "* **Data Visualization with IGV:**  Learners will utilize the `igv-notebook` library to visualize the aligned BAM files and methylation levels (bedGraph files) in the Integrative Genomics Viewer (IGV).  This includes loading the appropriate genome, alignment files and metadata within the IGV interface."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f95cc7b",
   "metadata": {},
   "source": [
    "## Prerequisites\n",
    "\n",
    "**1. Software:**\n",
    "\n",
    "* **Mamba (or conda):** A package manager for installing and managing software dependencies.  The notebook uses it to install the bioinformatics tools.\n",
    "* **Bioconda channel:** A conda channel containing bioinformatics tools.  Must be specified when installing the tools.\n",
    "* **FastQC:**  Quality control for high throughput sequence data.\n",
    "* **MultiQC:** Aggregates results from multiple FastQC reports into a single report.\n",
    "* **Bismark:** Aligns bisulfite-treated sequencing reads to a reference genome.\n",
    "* **Trim Galore!:**  Trims adapters and low-quality sequences from reads.\n",
    "* **Bedtools:** A suite of utilities for genomic analysis.\n",
    "* **Samtools:**  Utilities for manipulating SAM/BAM files (sequence alignment/mapping format).\n",
    "* **Metilene:**  Software for methylation analysis (although its use is not explicitly shown in this notebook excerpt).\n",
    "\n",
    "**2. APIs:**\n",
    "\n",
    "* **Google Cloud Storage API:** Required for downloading the example dataset and reference genome from the specified Google Cloud Storage bucket using `gsutil`.\n",
    "\n",
    "**Important Note:** The notebook contains placeholders like `<YOUR COMMAND HERE>`.  These need to be replaced with the appropriate commands based on the file paths and specific parameters required.  The help dropdowns provide examples; ensure those commands accurately match the setup and local system.  Remember to adapt file paths within commands to your local directory structure."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27af056b-5062-41e4-968c-d363e9420d7d",
   "metadata": {},
   "source": [
    "## Get Started"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d1cd3f7-38f0-4606-8b3c-0c5c6dc26fe2",
   "metadata": {},
   "source": [
    "Use mamba to install the tools needed for this analysis: fastqc, multiqc bismark, trim-galore, bedtools, samtools, and metilene. Remember to set the channel to bioconda in order to access the bioinformatics tools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a79de905-dc75-4d6f-a7a3-419069233bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae0368b-f4d2-47f3-899f-ab638a0ceb69",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`mamba install -y -c conda-forge -c bioconda fastqc multiqc bismark trim-galore bedtools samtools metilene` \n",
    "\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ccc3f4b-bdc8-40f3-be16-ade14562ad24",
   "metadata": {},
   "source": [
    "---\n",
    "### **Importing the example dataset**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374f0472-ace6-4bec-918e-a9cb6eebec2b",
   "metadata": {},
   "source": [
    "#### Set up the environment\n",
    "As before, we are going to set up the directory structure to keep everything organized. We will change `Tutorial_1` to `Tutorial_5`. Use the `mdkir` commands to create your directory structure below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c405c48-d374-4f54-a16a-6c61e75a7bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b2a90ef-606a-40f0-9d0b-d334f9d9b154",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`# Show current working directory`  \n",
    "`pwd`  \n",
    "`# Create directories`  \n",
    "`mkdir -p Tutorial_5`  \n",
    "`mkdir -p Tutorial_5/ref_genome`  \n",
    "`mkdir -p Tutorial_5/fastqc`  \n",
    "`mkdir -p Tutorial_5/trimmed`   \n",
    "`mkdir -p Tutorial_5/fastq`  \n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd74464d-d6c3-4388-822b-3ca3ebb93934",
   "metadata": {},
   "source": [
    "#### About the new dataset\n",
    "\n",
    "Our new dataset comes from a paper by [Hadad et al. Epigenetics Chromatin. 2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6781367/) that compares methylation changes in mice as they age and correlates those changes to gene expression changes. The data is available in SRA under the bioProject number [PRJNA523985](https://www.ncbi.nlm.nih.gov/sra/?term=PRJNA523985). The impact of methylation on aging, particularly in brain tissue, is of great research interest. The paper makes several comparisons, but to keep this tutorial simple we only pulled the samples of young and old female mice. We also subsamples the data to 10% of the reads on SRA, otherwise the data would be very large and would require long compute times and higher costs than would be appropriate for a tutorial.\n",
    "\n",
    "\n",
    "#### Download example dataset and reference genome from Google Cloud Storage bucket\n",
    "\n",
    "This WGBS example dataset was stored in a **Google Cloud Storage bucket**. You can use `gsutil` commands to access and manage the cloud storage. \n",
    "\n",
    "To download this example dataset, we need to get the fastq.gz files, the samplesheet (.tsv), and the CpG island annotation file (.bed) from the storage bucket to the Tutorial_5 directory we just created using `gsutil cp`. In the cell below, type the commands to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2d9fb4f-29f1-438d-bc32-5a667add501a",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74058795-6a9c-4dd1-8fb0-44d9d29899b3",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`# Download the example dataset from a Google Cloud Storage bucket`  \n",
    "`gsutil cp gs://nigms-sandbox/dna-methyl/*.fastq.gz Tutorial_5/fastq`  \n",
    "`# Download the samplesheet`  \n",
    "`gsutil cp gs://nigms-sandbox/dna-methyl/*.tsv Tutorial_5/`  \n",
    "`# Download the bed file`  \n",
    "`gsutil cp gs://nigms-sandbox/dna-methyl/*.bed Tutorial_5/`  \n",
    "\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f4f8d31-1035-4ec9-bcaf-1d34493d8b07",
   "metadata": {},
   "source": [
    "Next download the **reference genome** in the same way and place it in the `ref_genome` folder. The genome should be in FASTA format, and can be downloaded from NCBI or Ensemble websites. As before, we will only use the sequences from GRCm39 chromosome 6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83cf634-1bed-4fa4-9428-77616d696943",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0315ed0-4895-42b8-9c77-db215cc6d426",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`gsutil cp gs://nigms-sandbox/dna-methyl/Mus_musculus.GRCm39.dna.chromosome.6.fa Tutorial_5/ref_genome`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e531a87-97ae-4092-8a50-54170070306c",
   "metadata": {},
   "source": [
    "---\n",
    "## **Running Bismark pipeline**\n",
    "\n",
    "#### Step 1. FastQC \n",
    "\n",
    "As before, we start with [**FastQC**](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to evaluate the quality of the sequence data we got from SRA. Write a command below to run FastQC on the new data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "579e18ff-3d36-4e5f-a00f-a1bd013c5de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "218020d2-be9a-4a4c-b2c1-2c4117c47ba8",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`for file in Tutorial_5/fastq/*.gz; do fastqc -q -o Tutorial_5/fastqc \"${file}\"; done;`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a108da6-fb87-4050-b1de-476a4be641ea",
   "metadata": {},
   "source": [
    "#### Step 2. MultiQC (optional)\n",
    "Next, we will use __[MultiQC](https://multiqc.info/)__ to aggregate results from across the samples into a single report. As before, save the output in `Tutorial_5/multiqc`. Below, write the commands to create the report and then view it as an html file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c6cbb6d-a7f6-4826-b0cc-299699ac1d6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbf4d0e9-35c4-4656-b39d-320b0e644a45",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`# Run multiqc to summarize all the fastqc reports`  \n",
    "`multiqc -f -p  Tutorial_5/fastqc -o Tutorial_5/multiqc`  \n",
    "  \n",
    "`# View multiqc report`  \n",
    "`from IPython.display import IFrame`  \n",
    "`IFrame(src='Tutorial_5/multiqc/multiqc_report.html', width=1200, height=400)`  \n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cbdf7bb-58eb-412f-8a51-c48c2269da58",
   "metadata": {},
   "source": [
    "#### Step 3. Trim the adapters and low quality sequence reads\n",
    "\n",
    "Adapter sequences should be removed from reads because they interfere with downstream analyses, such as alignment of reads to a reference.\n",
    "\n",
    "As before, we will use __[Trim Galore!](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)__ for adaptor trimming and quality control. Using the Tutorial 1 trim_galore command as a template, write a command to run trim_galore on your reads. Use wildcards to match your fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d97717-e599-4fd7-a3a9-e955f0c83e26",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "549d11fa-3aa7-4e87-884c-d863bcd06501",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`trim_galore -j 4 --paired --illumina -o Tutorial_5/trimmed --fastqc Tutorial_5/fastq/SRR*.gz`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1070cc02-b431-449b-bfdf-8d28e0d0134c",
   "metadata": {},
   "source": [
    "#### Step 4. Genome preparation\n",
    "\n",
    "In order to align the bisulfite-treated reads to reference genome, the reference genome need to be converted as well. \n",
    "\n",
    "Write the command below to prepare the reference genome. Do we need to change anything? Or can we leave it the same as with Tutorial 1? What factors do you need to consider to determine this?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8d70b3-f28f-4418-95ec-41d4e22e6254",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c35ca2d-11c6-43e4-82b3-e7dfb73700c6",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "    Since we are using the same reference genome, the command stays the same. The only change we need to make is to the path where it is stored.\n",
    "\n",
    "    \n",
    "\n",
    "`bismark_genome_preparation Tutorial_5/ref_genome`\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56945ac-bde8-41dc-b8b5-6299032627fd",
   "metadata": {},
   "source": [
    "#### Step 5. Run Bismark (alignment)\n",
    "\n",
    "We next run Bismark to align the reads to the reference genome. If you recall from Tutorial 1, the usage for Bismark is:\n",
    "\n",
    "`bismark [options] --genome <genome_folder> {-1 <mates1> -2 <mates2> | <singles>}`\n",
    "\n",
    "In the command block below, write your adapted Bismark command for your new dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0e14cc-42b1-4362-ae06-0b9c4b788511",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2926ea11-37de-41bd-8a22-578712c627a6",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`for file_name in Tutorial_5/trimmed/*_R1_val_1.fq.gz; do \\\n",
    "    bismark --genome Tutorial_5/ref_genome/ -1 ${file_name} -2 ${file_name//_R1_val_1/_R2_val_2} -o Tutorial_5/bismark; \\\n",
    "done;`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61be513c-5bdf-42a9-a03b-ed41d771c120",
   "metadata": {},
   "source": [
    "#### Step 6. De-duplicate \n",
    "\n",
    "Recall from Tutorial 1 that a high level of duplication is more likely to indicate some kind of enrichment bias (eg PCR over amplification). To mitigate that potential bias, we will run a de-duplication process on our alignments.\n",
    " \n",
    "> <img src=\"images/1_complex_library.png\" width=\"500\"/>   \n",
    "\n",
    "In the command block below, write your `deduplicate_bismark` command with your new data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5146895c-8e09-493d-821a-8aa053d67f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d41236-75f9-4666-8558-97b6f268b85f",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`deduplicate_bismark -p --output_dir Tutorial_1/bismark --bam Tutorial_1/bismark/*.bam `\n",
    "\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e02055-79c7-4f77-9669-43fd62479ff3",
   "metadata": {},
   "source": [
    "#### Step 7. Get methylation ratios \n",
    "\n",
    "This step operates on Bismark alignment files and extracts the methylation call for every single C analyzed. Think about what command is used to do this. What input files do you need? Do you need to change any flags for this run? What directory do you need to write the output to? Write your command in the cell below try it out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3842a4f4-ba74-4ea8-a50c-86e3d1031e46",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c3cf35-4b09-40f4-8752-108bb9f04fb0",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`bismark_methylation_extractor --gzip --bedGraph --ignore_r2 2 -o Tutorial_1/bismark Tutorial_1/bismark/*deduplicated.bam`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9855179f-507f-4c04-8559-6b41ad4f9cdd",
   "metadata": {},
   "source": [
    "As a sanity check, look at the first few lines of a couple output files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439c1191-832d-483f-8cb4-dcefc1dd78f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "843098aa-2aab-4f39-8771-1204f5444d15",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`zcat Tutorial_5/bismark/SRX202087_R1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz | head -n 10`  \n",
    "`zcat Tutorial_5/bismark/SRX202087_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz | head -n 10`\n",
    "\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c868907a-2278-4c3c-8b49-4f734ad149ab",
   "metadata": {},
   "source": [
    "#### Step 8. Generate report and summary \n",
    "Finally, generate a report and summary of the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aea84f5c-c30a-456c-be38-6c7c02560093",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0ff0de-9611-49af-bd5e-9adec62473f6",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "```\n",
    "cd Tutorial_5/bismark && bismark2report\n",
    "cd Tutorial_5/bismark && bismark2summary  \n",
    "# Display the final summary from bismark workflow, including all 4 samples\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"Tutorial_5/bismark/bismark_summary_report.txt\", sep='\\t')\n",
    "df\n",
    "```\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "269514d7-5ead-4176-ba11-b211041b75a9",
   "metadata": {},
   "source": [
    "#### Important output files for downstream analysis\n",
    "\n",
    "The Bismark methylation extractor (step 7) can output files in `.bedGraph` and `.cov` format with the methylation levels (percentage of methylated cytosine) at each position. These files can be used as input for downstream analysis, such as identification of differentially methylated regions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb468d2f-9023-498e-918e-3053d33efdc9",
   "metadata": {},
   "source": [
    "---\n",
    "## **Visualization using IGV (Integrative Genomics Viewer)**\n",
    "\n",
    "Let's look at our recently analyzed data with Integrative Genomics Viewer. Remember the three basic steps of using IGV as you load your new data.\n",
    "\n",
    "#### Basic usage\n",
    "- Select reference genome - IGV hosts dozens of genomes and you can load other genomes too\n",
    "- Load data tracks\n",
    "- Navigate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "351d564e-e539-4e41-84e0-e6a30db62ea9",
   "metadata": {},
   "source": [
    "#### Install `igv-notebook` \n",
    "\n",
    "Install igv-notebook with pip."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f078cf79-8a95-4eef-b94d-f871ad625aa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bd17fb-da30-4ff9-8a49-530acb070736",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`pip install igv-notebook`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7071182-60b9-42f6-8ea9-2beffb2e47f1",
   "metadata": {},
   "source": [
    "Create a browser and load your reference genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "972fd6de-ae71-4168-86a7-1952c1717565",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND BELOW>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10e26794-03bc-461a-8fac-66182ff58325",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`import igv_notebook`  \n",
    "`igv_notebook.init()`  \n",
    "\n",
    "`b = igv_notebook.Browser(`  \n",
    "`    {`  \n",
    "`        \"genome\": \"mm39\",`  \n",
    "`        \"locus\": \"chr6:4,733,000-4,749,000\"`  \n",
    "`    }`  \n",
    "`)`  \n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2210a9a3-050d-4274-bbde-e8584cc512fa",
   "metadata": {},
   "source": [
    "Next, load the data tracks. This requires a sorted and indexed bam files. Write the command below to do that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30bbae5d-0b4c-4fbb-9842-cb03eb0db181",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11507a79-9193-4a09-b41f-d83d00f94f17",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`for i in Tutorial_5/bismark/*deduplicated.bam; do samtools sort $i -o $i.sort.bam; samtools index $i.sort.bam;done`  \n",
    "`for f in Tutorial_5/bismark/*sort.bam* ; do mv \"$f\" \"$(echo \"$f\" | sed s/_R1_val_1_bismark_bt2_pe.deduplicated.bam//)\"; done`  \n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "156e533b-b111-4788-883b-ecd5811619b3",
   "metadata": {},
   "source": [
    "#### Load and view the alignment file (and its index) to igv-notebook.\n",
    "\n",
    "Start a new browser \"b2\", and load one of sample alignment .bam and its index file .bai into the browser:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c520603-3fc7-4b53-9836-b3d849b032b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d96bca4-786f-4618-982e-0e5bed7fce1a",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`b2 = igv_notebook.Browser(`  \n",
    "`    {`  \n",
    "`       \"genome\": \"mm39\",`  \n",
    "`       \"locus\": \"chr6:4,733,000-4,749,000\"`  \n",
    "`    }`  \n",
    "`)`  \n",
    "`b2.load_track({`  \n",
    "`    \"name\": \"SRX271141\",` \n",
    "`    \"path\": \"Tutorial_1/bismark/NAME.sort.bam\",`  \n",
    "`    \"indexPath\": \"Tutorial_1/bismark/NAME.sort.bam.bai\",` \n",
    "`    \"format\": \"bam\",`  \n",
    "`    \"type\": \"alignment\",`  \n",
    "`})`  \n",
    "\n",
    "</details>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da63305e-8015-4639-bb7f-4b04a0504105",
   "metadata": {},
   "source": [
    "<div class=>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ac59d7-0997-4208-870f-08c10e1ff92c",
   "metadata": {},
   "source": [
    "#### View the methylation levels (.bedgraph files)\n",
    "\n",
    "Next, load the bedgraph files to view the methylation percentage at each position. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599856a9-891c-426b-bd43-4e6afbb69a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "<YOUR COMMAND HERE>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4edd38a4-de64-4149-99e1-6e2865479f08",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`b3 = igv_notebook.Browser({\"genome\": \"mm39\",\"locus\": \"chr6:4,500,000-5,500,000\"})`   \n",
    "  \n",
    "`b3.load_track({ \"name\": \"SRX271141_bed\",\"path\": \"Tutorial_1/bismark/SRX271141_R1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\",\"mode\":\"bisulfite\" ,\"color\": \"orange\"})`  \n",
    "      \n",
    "`b3.load_track({` \n",
    "`    \"name\": \"SRX202087_bed\",` \n",
    "`    \"path\": \"Tutorial_1/bismark/SRX202087_R1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\",` \n",
    "`    \"mode\": \"bisulfite\" ,` \n",
    "`    \"color\": \"orange\"` \n",
    "`})` \n",
    "  \n",
    "`b3.load_track({` \n",
    "`    \"name\": \"SRX271142_bed\",` \n",
    "`    \"path\": \"Tutorial_1/bismark/SRX271142_R1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\",` \n",
    "`    \"mode\": \"bisulfite\" ,` \n",
    "`    \"color\": \"blue\"` \n",
    "`})` \n",
    "    \n",
    "    \n",
    "`b3.load_track({` \n",
    "`    \"name\": \"SRX202088_bed\",` \n",
    "`    \"path\": \"Tutorial_1/bismark/SRX202088_R1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\",` \n",
    "`    \"mode\": \"bisulfite\" ,` \n",
    "`    \"color\": \"blue\"` \n",
    "`})` \n",
    "\n",
    "\n",
    "</details>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ebbdd40-85c5-4629-959a-fc91fa756665",
   "metadata": {},
   "source": [
    "Finally, zoom into a specific region to have a closer look."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a8eb0df-b980-4b06-aa72-e1ee6aaede73",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary>Click for help</summary>\n",
    "\n",
    "\n",
    "`# Zoom in to see a specific region in the above browser`  \n",
    "`type your region in the format chr<number>:start-end`  \n",
    "`b3.search('<region_here>')`\n",
    "\n",
    "\n",
    "</details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf2214cf",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "\n",
    "This Jupyter Notebook demonstrated the application of a WGBS data analysis module to a new dataset from Hadad et al. (2019), detailing the process of adapting the pipeline for novel data. We successfully reproduced the Bismark pipeline steps, including quality control with FastQC and MultiQC, adapter trimming with Trim Galore!, genome preparation, Bismark alignment, deduplication, methylation extraction, and report generation. The final methylation data was visualized using IGV-notebook, showcasing the workflow's effectiveness in analyzing WGBS data and providing a template for future analyses. While the tutorial focused on a specific dataset, the adaptability of the module is highlighted by its successful implementation with the new data, suggesting its potential for broad use in WGBS research."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d1baae",
   "metadata": {},
   "source": [
    "## Clean Up\n",
    "**Don't forget:** After you finish running the notebook, stop it in Vertex AI Workbench  to avoid accumulating costs."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
